subprojects/plugins/src/main/java/org/gradle/api/plugins/JavaPlugin.java

```diff
@@ -198,4 +204,25 @@ public Configuration getCompileDependencies() {
             return convention.getProject().getConfigurations().getByName(JavaPlugin.COMPILE_CONFIGURATION_NAME);
         }
     }
+
+    private class SourceSetOutputCleanUpRegistrationListener implements ProjectEvaluationListener {
+
+        private final SourceSet main;
+        private final SourceSet test;
+        private final BuildOutputCleanupRegistry buildOutputCleanupRegistry;
+
+        public SourceSetOutputCleanUpRegistrationListener(SourceSet main, SourceSet test, BuildOutputCleanupRegistry buildOutputCleanupRegistry) {
+            this.main = main;
+            this.test = test;
+            this.buildOutputCleanupRegistry = buildOutputCleanupRegistry;
+        }
+
+        @Override
+        public void beforeEvaluate(Project project) {}
+
+        @Override
+        public void afterEvaluate(Project project, ProjectState state) {
+            buildOutputCleanupRegistry.registerOutputs(main.getOutput().getClassesDir(), main.getOutput().getResourcesDir(), test.getOutput().getClassesDir(), test.getOutput().getResourcesDir());
```

Comment:
I removed the unused method that takes `List<File>` in favor of `FileCollection`. Good idea.